org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/folding/FoldingRangeService.java

```diff
@@ -0,0 +1,52 @@
+/*******************************************************************************
+ * Copyright (c) 2021 TypeFox GmbH (http://www.typefox.io) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server.folding;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.lsp4j.FoldingRange;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.xtext.ide.editor.folding.IFoldingRangeProvider;
+import org.eclipse.xtext.ide.server.Document;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * @author Mark Sujew - Initial contribution and API
+ */
+@Singleton
+public class FoldingRangeService {
+
+	@Inject
+	private IFoldingRangeProvider foldingRangeProvider;
+
+	public List<FoldingRange> createFoldingRanges(Document document, XtextResource resource,
+			CancelIndicator cancelIndicator) {
+		return foldingRangeProvider.getFoldingRanges(resource, cancelIndicator).stream()
+				.map(range -> toFoldingRange(document, range))
+				.filter(range -> range.getStartLine() < range.getEndLine()).collect(Collectors.toList());
+	}
+
+	protected FoldingRange toFoldingRange(Document document, org.eclipse.xtext.ide.editor.folding.FoldingRange range) {
+		int offset = range.getOffset();
+		int length = range.getLength();
+		int endOffset = offset + length;
+		Position start = document.getPosition(offset);
+		Position end = document.getPosition(endOffset);
+		FoldingRange result = new FoldingRange(start.getLine(), end.getLine());
+		result.setStartCharacter(offset);
```

Comment:
I may be wrong here, but I understand the spec as these chars being columns in the folded line. Can you confirm that these are supposed to be relative to the document?

Reply:
Yes, you're right. I somehow got this wrong.

Reply:
Wait, on second rereading, it sounds different again: "If not defined, defaults to the length of the start line." Is the *length* here meant as the length of the whole document string or as the length of the line?

Reply:
To me, `defaults to the length of the start line` sounds like the length of the (start) line ;)

Reply:
> defaults to the length of the start line
I understand it as the length of the line the `FoldingRange.startLine` variable refers to.

Reply:
I've looked into different language server implementations ([java](https://github.com/eclipse/eclipse.jdt.ls/blob/3a52306c1eda1c409c68f5d80bb0d64e518c2233/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FoldingRangeHandler.java#L139), [ts](https://github.com/theia-ide/typescript-language-server/blob/3a8ea0f04bde047e2f60b9f68ef7816e3a41aa10/server/src/lsp-server.ts#L887-L912), [R](https://github.com/REditorSupport/languageserver/blob/21d796ba4167ab457cace13b6096121adde3e6d9/R/folding.R#L73-L77)) and neither actually use the `startCharacter` or `endCharacter` properties. So I don't have any reference how everything is actually meant.

Is there an actual benefit of implementing it with this PR? Not even vscode can actually handle these properties (returns `lineFoldingsOnly=true`)

Reply:
would you please be so kind and ask it as a clarifining question issue in the protocol github repo?

Reply:
Alright https://github.com/microsoft/language-server-protocol/issues/1198

Reply:
> Is there an actual benefit of implementing it

When in doubt leave it out

Reply:
I removed the start- and endCharacter setters. I've created #1675 as a reminder.

@miklossy It may be necessary to update your tests, as the ordering of the FoldingRanges may have changed as well (using a `SortedSet` to hold them now)

Reply:
ok, thanks for the reminder. I will try it out now.